pubtools/_pulp/tasks/copy_repo.py

```diff
@@ -0,0 +1,324 @@
+            if push_item:
+                out.append(push_item)
+        return out
+
+    def push_item_for_unit(self, unit):
+        for unit_type, fn in [
+            (ModulemdUnit, self.push_item_for_modulemd),
+            (RpmUnit, self.push_item_for_rpm),
+            (FileUnit, self.push_item_for_file),
+        ]:
+            if isinstance(unit, unit_type):
+                return fn(unit)
+
+    def push_item_for_modulemd(self, unit):
+        out = {}
+        out["state"] = "PUSHED"
+        out["origin"] = "pulp"
+
+        # Note: N:S:V:C:A format here is kept even if some part
+        # of the data is missing (never expected to happen).
+        # For example, if C was missing, you'll get N:S:V::A
+        # so the arch part can't be misinterpreted as context.
+        nsvca = ":".join(
+            [unit.name, unit.stream, str(unit.version), unit.context, unit.arch]
+        )
+
+        out["name"] = nsvca
+
+        return ModuleMdPushItem(**out)
+
+    def push_item_for_rpm(self, unit):
+        out = {}
+
+        out["state"] = "PUSHED"
+        out["origin"] = "pulp"
+
+        filename_parts = [
+            unit.name,
+            "-",
+            unit.version,
+            "-",
+            unit.release,
+            ".",
+            unit.arch,
+            ".rpm",
+        ]
+        out["name"] = "".join(filename_parts)
+
+        # Note: in practice we don't necessarily expect to get all of these
+        # attributes, as after a delete the server will only provide those
+        # which make up the unit key. We still copy them anyway (even if
+        # values are None) in case this is improved some day.
+        out["sha256sum"] = unit.sha256sum
+        out["md5sum"] = unit.md5sum
+        out["signing_key"] = unit.signing_key
+
+        return RpmPushItem(**out)
+
+    def push_item_for_file(self, unit):
+        out = {}
+
+        out["state"] = "PUSHED"
+        out["origin"] = "pulp"
+        out["name"] = unit.path
+        out["sha256sum"] = unit.sha256sum
+
+        return FilePushItem(**out)
+
+    @step("Publish")
+    def publish(self, repo_fs):
+        return [
+            f_flat_map(f, lambda r: r.publish(PublishOptions(clean=True)))
+            for f in repo_fs
+        ]
+
+    @step("Flush UD cache")
+    def flush_ud(self, repos):
+        client = self.udcache_client
+        if not client:
+            LOG.info("UD cache flush is not enabled.")
+            return []
+
+        out = []
+        for repo in repos:
+            out.append(client.flush_repo(repo.id))
+            if repo.eng_product_id:
+                out.append(client.flush_product(repo.eng_product_id))
+
+        return out
+
+    def run(self):
+        to_await = []
+
+        # Start copying repos.
+        repos_to_flush = []
+        repo_copies_fs = []
+        for pair in self.repo_pairs:
+            LOG.info("Copying repo %s to %s", pair[0], pair[1])
+            src_repo, dest_repo = self.get_repos(pair)
```

Comment:
Relating to my earlier comment about what `get_repos` is supposed to accept:

The code here is why I'm saying it looks like it's supposed to accept just one pair, and not a list of pairs like a comment within the method suggests.

However, I think you probably should change it so it does accept a list of pairs.  I think the main point of that code was to verify that the user had provided valid IDs for all of their repos before we start doing any real work.  The way this loop is set up, it won't accomplish that now. It could be fixed if get_repos indeed accepted a list of pairs of repo IDs, `list[tuple[str,str]]` and returned a list of pairs of repos `list[tuple[Repository,Repository]]`.

For example if the user passes "src1,dst1 src2,dst2 src3,dst3" and src3 doesn't exist, we ideally want the task to quickly determine that and exit with an error.  Right now it looks like it would start the copies between src1/dst1 and src2/dst2 before crashing on src3.